random_seed.py

```diff
@@ -187,32 +181,60 @@ def generate(self, world_graph = None):
                         StarSpirits.KLEVAR,
                         StarSpirits.KALMAR,
                     ]
-                    chosen_spirits = []
                     for _ in range(self.rando_settings.starway_spirits_needed_count):
                         rnd_spirit = random.randint(0, len(all_spirits) - 1)
                         chosen_spirits.append(all_spirits.pop(rnd_spirit))
                     encoded_spirits = 0
                     for spirit in chosen_spirits:
                         encoded_spirits = encoded_spirits | (1 << (spirit - 1))
                     self.rando_settings.starway_spirits_needed_encoded = encoded_spirits
-                    modified_world_graph = get_specific_spirits(
-                        modified_world_graph,
-                        chosen_spirits
-                    )
 
-                    if self.rando_settings.limit_chapter_logic:
-                        modified_world_graph = get_limited_chapter_logic(
-                            modified_world_graph,
-                            chosen_spirits,
-                            self.rando_settings.gear_shuffle_mode
-                        )
                     chosen_spirits.sort()
                     if self.spoilerlog_additions.get("required_spirits") is None:
                         self.spoilerlog_additions["required_spirits"] = []
                     self.spoilerlog_additions["required_spirits"].extend(chosen_spirits)
-                else:
-                    self.rando_settings.require_specific_spirits = False
-                    self.rando_settings.starway_spirits_needed_encoded = 0xFF
+
+                if (   self.rando_settings.starbeam_spirits_needed > 0
+                    or self.rando_settings.starbeam_powerstars_needed > 0
+                ):
+                    modified_world_graph = set_starbeam_requirements(
+                        world_graph=modified_world_graph,
+                        spirits_needed=self.rando_settings.starbeam_spirits_needed,
+                        powerstars_needed=self.rando_settings.starbeam_powerstars_needed
+                    )
+
+                entrance_changes, modified_world_graph = set_starway_requirements(
+                    world_graph=modified_world_graph,
+                    spirits_needed=self.rando_settings.starway_spirits_needed_count,
+                    specific_spirits=chosen_spirits,
+                    powerstars_needed=( # don't expect all, but also don't bottleneck
+                        self.rando_settings.star_hunt_total
+                      - int(  (  self.rando_settings.star_hunt_total
```

Comment:
I don't understand the math here

Reply:
If we just used `starhunt_total`, then in a, say, "30 out of 50" seed we'd logically require all 50 power stars before star way.
So even though you only really need 30, none of the other 20 stars could be in chapter 8.
If instead we used `starway_powerstars_needed`, then we could place the remaining 20 into chapter 8, but if you get unlucky then only exactly the 30 you need are even available before star way, meaning there's no buffer.
So the math here calculates the buffer (placed minus needed), divides it by 2 and adds it onto needed.
So it'd be 50 - ((50 - 30) / 2) = 40. By logically requiring the middle between needed and placed you have both a buffer and stars can be placed in chapter 8